### DISCUSSION TITLE
How to handle different data for different users in Wave App?

### DISCUSSION THREAD

**Discussion by n-srinidhi** [h2oai/wave#2234] [Q&A]
Hi!I have built a wave application in which different users can login with their credentials and view data relevant to them. I have handled the login authentication bit independently via background apis and the authentication and first time data cards creation works fine. However when the updates are received by Wave server, The same update is reflected on the dashboard of two different users,on different client tabs I have been reading the documentation trying to understand `q.client` vs `q.user` variables. How do i update user wise? How does Wave understand which browser client of which user to be updated? I used `mode=multicast` since I want all browser tabs for a particular user logged in to show updated data. However, The user separation bit isn't done properly. Can you help me understand how to fix this?

Note: I have a backend server which pushed updates to the front end via web sockets. I am storing a websocket client at `q.client` level as `q.client.engine_client` so this socket receives updated data. I am handling receiving data etc via callback based web socket library.

**Comment by marek-mihok** [comment]
> How does Wave understand which browser client of which user to be updated?

`q.user` is a state shared across all open tabs (clients) within a single browser window.

Are you trying to have multiple users logged in inside separate tabs within single browser window? Do I understand it correctly?

**Reply by n-srinidhi** [reply]
Yes!
Will it work across different browser windows then?

**Reply by marek-mihok** [reply]
> Will it work across different browser windows then?

Hard to tell without seeing your code, but it should work. Please let me know if you try.

Is there a reason for having different users logged in inside different tabs within the same window by the way?

**Comment by n-srinidhi** [comment]
[sample_wave_app_multiple_clients.zip](https://github.com/h2oai/wave/files/13958486/sample_wave_app_multiple_clients.zip)

HI!Im sharing a sample repo to illustrate whats happening. I'll try executing it in different browser windows to check if it works. 
Since this is being deployed in production, It is a requirement to have multiple User IDs logged in via a single browser windows and view their respective data parallely.

**Reply by n-srinidhi** [reply]
I tried it with different browser windows and over time the updates were mixed on different windows as well.
